Mathlib/Algebra/Star/Positive.lean

```diff
@@ -0,0 +1,183 @@
+/-
+Copyright (c) 2023 Jireh Loreaux. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jireh Loreaux
+-/
+import Mathlib.Algebra.Order.Nonneg.Ring
+import Mathlib.Algebra.Star.Order
+import Mathlib.Algebra.Star.SelfAdjoint
+import Mathlib.GroupTheory.Submonoid.Operations
+
+/-!
+# Positive elements in a star ordered ring
+
+This file defines the `AddSubmonoid` of nonegative elements `StarOrderedRing.positive R` in a
+star ordered ring `R` with carrier `{ x : R | 0 ≤ x }` (in fact, the definition only requires
+`OrderedAddCommMonoid R` to allow for other use caess, but it is primarily intended for
+`StarOrderedRing R`). Equivalently, this is the `AddSubmonoid` generated by the elements of the
+form `star s * s`. Even though this technically corresponds to the nonnegative elements, the
+"positive" terminology is standard throughout the literature (at least in C⋆-algebra theory and
+operator theory), so we choose to prefer that.
+
+The advantage of using this over other definitions of positivity is that it allows us to unify
+several different concepts under a single umbrella. For example, one might be tempted to consider
+the collection of bounded linear operators on an Hilbert space and define positivity as
+`∀ x, 0 ≤ ⟪T x, x⟫`, but this doesn't generalize nicely to C⋆-algebras. Alternatively, one could
+define positivity as those self-adjoint elements with nonnegative spectrum, but this requires the
+algebra to be unital. Our definition herein conveniently sidesteps these issues.
+
+We place everything in the `StarOrderedRing` namespace in order to avoid collisions or confusion
+with other parts of the library.
+
+## Main definitions
+
+* `StarOrderedRing.positive R`: The additive submonoid of a star ordered ring `R` generated by
+  elements of the form `star s * s` for `s : R`.
+* `StarOrderedRing.positive.selfAdjoint : positive R →+ selfAdjoint R`: the identity map as an
+  `AddMonoidHom`.
+* `StarOrderedRing.positive.instAddMonoidWithOne`: Makes `positive R` into an `AddMonoidWithOne`
+* `StarOrderedRing.positive.instCanonicallyOrderedAddMonoid`: Makes `positive R` into a
+  `CanonicallyOrderedAddMonoid` which avoids duplication of some lemmas (e.g., `zero_le`)
+
+-/
+
+universe u
+
+variable {R : Type u}
+
+namespace StarOrderedRing
+
+/-- The positive elements of a star ordered ring, as an additive submonoid.
+
+Note that the type class assumptions for `StarOrderedRing.positive` are actually significantly
+weaker (just `OrderedAddCommMonoid`) so that it can be used in as many contexts as possible, but
+the primary intended use case is when `R` is a `StarOrderedRing`.  -/
+def positive (R : Type u) [OrderedAddCommMonoid R] : AddSubmonoid R where
```

Comment:
I'd argue this belongs in nonneg/ring, as it can be used to golf some instances there

Reply:
I've put this (and one for `Subsemiring`) in the appropriate subobject files; I didn't place them in `Algebra.Order.Nonneg.Ring` in order to avoid import creep. Now `StarOrderedRing.positive` is just an alias for `AddSubmonoid.nonneg`, but we keep a separate name because we prefer it and also because we will place many instances on it.